nep-dapi.mediawiki

```diff
@@ -0,0 +1,586 @@
+</pre>
+
+'''Input arguments'''
+
+None
+
+'''Success return value'''
+
+object - Provider information defined by the interface below:
+<pre>
+interface Provider {
+  name: string;
+  website: string;
+  version: string;
+  compatibility: string[];
+}
+</pre>
+
+The compatibility field will return a list of strings that corresponds to NEPs that the dAPI conforms to. For example:
+<pre>
+compatibility: [
+    'NEP-14',
+    'NEP-23',
+    'NEP-29'
+]
+</pre>
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_DENIED`
+
+'''Example'''
+
+<pre>
+getProvider()
+.then((provider: Provider) => {
+  const {
+    name,
+    website,
+    version,
+    compatibility,
+  } = provider;
+
+  console.log('Provider name: ' + name);
+  console.log('Provider website: ' + website);
+  console.log('Provider dAPI version: ' + version);
+  console.log('Provider dAPI compatibility: ' + JSON.stringify(compatibility));
+})
+.catch((err: string) => {
+  switch(err) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp.');
+      break;
+  }
+});
+</pre>
+
+====getAccount====
+Return the Account that is currently connected to the dApp.
+
+'''Method Interface'''
+
+<pre>
+function getAccount(): Promise<{ address: string, publicKey: string }>
+</pre>
+
+'''Input arguments'''
+
+None
+
+'''Success return value'''
+
+address: string - Address of the connected account
+
+publicKey: string - Address of the connected account
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_DENIED`
+
+'''Example'''
+
+<pre>
+getAccount()
+.then((account: Account) => {
+  const {
+    address,
+    publicKey,
+  } = account;
+
+  console.log('Provider address: ' + address);
+  console.log('Provider public key: ' + publicKey);
+})
+.catch((err: string) => {
+  switch(err) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_DENIED:
+      console.log('The user rejected the request to connect with your dApp');
+      break;
+  }
+});
+</pre>
+
+
+
+====getBalance====
+Return balance of a specific asset for the given account.
+
+'''Method Interface'''
+
+<pre>
+function getBalance({ address: string, asset: string }): Promise<{ address: string, asset: string, amount: string }}>
+</pre>
+
+'''Input arguments'''
+
+address: string - The asset symbol the check the balance for the connected account
+
+asset: string - The asset symbol the check the balance for the connected account
+
+
+'''Success return value'''
+
+address: string - Address of the connected account
+
+asset: string - Asset symbol for the requested balance
+
+amount: string - Parsed balance of the requested asset
+
+
+'''Error return value'''
+
+string - `NO_PROVIDER`|`CONNECTION_DENIED`
+
+'''Example'''
+
+<pre>
+getBalance({
+  address: 'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru',
+  asset: 'NKN'
```

Comment:
asset should be a hash. (we could potentially create a list of hashes for better developer experience)
```
const ASSETS = {
  NEO: ...
  GAS: ...
}
```

Reply:
@nickfujita that is not good, the client implementing the call should know what is the final result of it. So different clients could call different contracts ..

Reply:
Could you please elaborate? In what case would for example the NEX wallet extension have a different script hash for NKN than NOS?

Reply:
Low probability of happening but it could, for example there was two ONT hashes once in the past.

Reply:
Even if this is the case, wouldn't there be greater consistency if the handful of wallet providers were responsible for managing these mappings rather than each and every dApp that wants to build on NEO? If any wallet that a user trusts with their funds has the wrong script hash for any NEP5 token, isn't there a a larger issue at hand?

Even in the case of NOS, the values are described as script hash values, but they are seemingly being sourced back from the repo itself, rather than by the dApp developer. Since the wallet provider will want to provide the dApp developer with as much information as possible to make their development experience as simple as possible, why not just cut out this step?
https://docs.nos.io/docs/nos-client/api.html#assets